File: reconstructItinerary/reconstructItinerary.cpp
// clang++ reconstructItinerary.cpp -std=c++11

#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>
using namespace std;

class Solution {
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {
        vector<string> res;
        unordered_map<string, multiset<string> > neighbors;
        for (auto& ticket : tickets) {
            neighbors[ticket.first].insert(ticket.second);
        }
        dfs("JFK", neighbors, res);
        reverse(res.begin(), res.end());
        return res;
    }
    void dfs(string start, unordered_map<string, multiset<string> >& neighbors, vector<string>& res) {
        while (!neighbors[start].empty()) {
            multiset<string>::iterator it = neighbors[start].begin();
            string next = *it;
            neighbors[start].erase(it);
            dfs(next, neighbors, res);
        }
        res.push_back(start);
    }
};

int main()
{
    Solution s;
    vector<pair<string, string> > tickets = {{"JFK", "SFO"}, {"JFK", "ATL"}, {"SFO", "ATL"}, {"ATL", "JFK"}, {"ATL", "SFO"}};
    vector<string> res = s.findItinerary(tickets);
    for (string& ss : res)
        cout << ss << " ";
    cout << endl;
    return 0;
}

File: reconstructItinerary/reconstructItinerary.java
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        List<String> itinerary = new LinkedList<>();
        Map<String, PriorityQueue<String> > graph = new HashMap<>();
        for (List<String> ticket : tickets) {
            if (!graph.containsKey(ticket.get(0))) {
                graph.put(ticket.get(0), new PriorityQueue<String>());
            }
            graph.get(ticket.get(0)).add(ticket.get(1));
        }
        dfs(graph, itinerary, "JFK");
        return itinerary;
    }
    public void dfs(Map<String, PriorityQueue<String> > graph, List<String> itinerary, String startNode) {
        while (graph.containsKey(startNode) && !graph.get(startNode).isEmpty()) {
            dfs(graph, itinerary, graph.get(startNode).poll());
        }
        itinerary.add(0, startNode);
    }
}